Method moveGroup to move one group under another group

balcirakpeter opened the pull request:
Created method moveGroup to move one group structure under another group in same vo
Created exception GroupMoveNotAllowed for that method
Created tests to check if method works corectly
Modified method updateGroup to allsow update parentGroupId if group was moved

zlamalp commented on perun-core/src/test/java/cz/metacentrum/perun/core/entry/GroupsManagerEntryIntegrationTest.java:
```diff
@@ -2088,6 +2074,152 @@ public void getAssignedGroupsToFacility() throws Exception {
 
+	@Test (expected = GroupMoveNotAllowedException.class)
+	public void testMoveGroupTodiffVo() throws Exception {
+		System.out.println("\ntest moveGroup() to different vo");
```
We usually print test name like this: ``System.out.println(CLASS_NAME + "testMoveGroupTodiffVo");`` Please fix it in a rest of tests.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java:
```diff
@@ -232,6 +232,18 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
 			}
 			dbGroup.setDescription(group.getDescription());
 		}
+
```
I would prefer implementing own method, instead of using updateGroup(), since some people use our API directly and they could create inconsistency by changing ``parentGroupId``, but not doing the rest of logic.

zlamalp replied:
Regarding future necessity of audit message about moving group, you should probably create ``updateParentGroupId()`` in BL layer too, where audit message will be logged in a future. Then use BL version from ``moveGroup()``

balcirakpeter commented:
I've created new method for updating parentGroupId, changed test names and print.
I've also added check to group cycle in method moveGroup and test if that works correctly, removed wild cards for exceptions in GroupsmanagerBl and GroupsManagerEntryIntegrationTest.

stavamichal commented:
What about audit messages? How Engine, LDAPc or other connectors get info about this change? How does they process it?

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java:
```diff
@@ -258,6 +259,31 @@ public Group updateGroupName(PerunSession sess, Group group) throws InternalErro
 	}
 
+	public Group updateParentGroupId(PerunSession sess, Group group) throws InternalErrorException {
+		Utils.notNull(group.getName(), "group.getName()");
```
Is this check necessary ? Technically, name should always exists, but we do not check other params and you do not work with group name in this method.

balcirakpeter commented:
Method updateParentGroupId() created also in BL layer
Check for group name in updateParentGroupId() removed

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/api/GroupsManager.java:
```diff
@@ -902,4 +902,17 @@
+	 * @throws GroupNotExistsException
+	 * @throws PrivilegeException
+	 */
+	void moveGroup(PerunSession sess, Group destinationGroup, Group movingGroup) throws InternalErrorException, GroupNotExistsException, PrivilegeException, GroupMoveNotAllowedException;
```
We should support the case "make group top-level", in such case ``destinationGroup`` would be null. Please allow it in a method implementation and update corresponding check before throwing ``GroupMoveNotAllowedException``.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,94 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+		movingGroup.setParentGroupId(destinationGroup.getId());
+		movingGroup.setName(destinationGroup.getName()+":"+movingGroup.getShortName());
+		//we have to update group name in database
+		getGroupsManagerImpl().updateGroupName(sess, movingGroup);
```
I believe, that in BL of normal updateGroup() or in DB ? we make sure, that on each hierarchy level, group name is unique. By moving group we could create two groups with same name on same level. Please check corresponding code to make sure this doesn't happen. Thank you.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,94 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+
+	public void moveGroup(PerunSession sess, Group destinationGroup, Group movingGroup) throws InternalErrorException, GroupMoveNotAllowedException {
+
+		// check if moving group or destination group is null
```
Please make sure, that ``destinationGroup`` is never ``members`` group of a VO, since we do not allow subgroups for members groups. We have name constant for this somewhere. Thank you.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,94 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+		//Save old parent id for moving indirect members
+		int oldParentGroupId = movingGroup.getParentGroupId();
+		// We can move whole moving group tree under destination group
+		movingGroup.setParentGroupId(destinationGroup.getId());
```
In a case of new top-level move, set ``parentGroupId`` to ``null``.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,94 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+		int oldParentGroupId = movingGroup.getParentGroupId();
+		// We can move whole moving group tree under destination group
+		movingGroup.setParentGroupId(destinationGroup.getId());
+		movingGroup.setName(destinationGroup.getName()+":"+movingGroup.getShortName());
```
In a case of new top-level move, just use ``movingGroup.getShortName()``

balcirakpeter commented:
Added check if destination group is members group
Added check if group move would create conflict with group names
Added logs for audit
Implementation changed to allow move group also as top level group
Added new tests

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,178 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+						throw new InternalErrorException("Parent group does not exists for group " + gr);
+					}
+					//we have to update each subGroup name in database
+					getGroupsManagerImpl().updateGroupName(sess, gr);
```
You are missing audit message about group update ``getPerunBl().getAuditer().log(sess, "{} updated.", gr);`` after this line, since all subgroups were changed, like in a case of moving group to top-level.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,178 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+		if (movingGroup == null) {
+			throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " cannot be null.");
+		}
+
```
Please add check, that "movingGroup" is NOT "Members" group too, like you do for destinationGroup.

balcirakpeter commented:
moveGroup() added to RPC layer
added chcek if movingGroup is membersGroup
added test for that check
added audit message about group update

zlamalp commented on perun-base/src/main/java/cz/metacentrum/perun/core/api/exceptions/GroupMoveNotAllowedException.java:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @author Peter Balcirak
+ */
+public class GroupMoveNotAllowedException extends PerunException {
```
Can you please update this exception to contain ``movingGroup`` and ``destinationGroup`` objects ? I can later use them in GUI to display nice message and group names. Right now exception message contains only serialized object.

balcirakpeter commented:
Added movingGroup and destinationGroup objects to GroupMoveNotAllowedException.

zlamalp commented on perun-base/src/main/java/cz/metacentrum/perun/core/api/exceptions/GroupMoveNotAllowedException.java:
```diff
@@ -0,0 +1,35 @@
+		super(message);
+		this.movingGroup = movingGroup;
+		this.destinationGroup = destinationGroup;
+	}
```
Please add also getters for ``movingGroup`` and ``destinationGroup``.

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java:
```diff
@@ -408,6 +408,187 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+			}
+
+			//Save old parent id for moving indirect members
+			int oldParentGroupId = movingGroup.getParentGroupId();
```
This will throw NullPointerException (I tested it) when group has no parent group, because then this attribute is null and you can't save null to int.

balcirakpeter commented:
I've moved removing/adding indirect members before the process of moving group, so we won't need that Integer. Instead of that we are removing indirect members just in case movingGroup is not top level group.
I've also created some tests for that so I hope its working correctly.

zlamalp commented on perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/GroupsManagerMethod.java:
```diff
@@ -173,6 +173,28 @@ public Group call(ApiCaller ac, Deserializer parms) throws PerunException {
 	},
 
+	/*#
+	 * Create union of two groups, where "operandGroup" is technically set as subgroup of "resultGroup".
```
This RPC javadoc is copy/pasted, please update it to represent moveGroup() logic.

zlamalp commented on perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/GroupsManagerMethod.java:
```diff
@@ -173,6 +173,28 @@ public Group call(ApiCaller ac, Deserializer parms) throws PerunException {
+			ac.stateChangingCheck();
+
+			ac.getGroupsManager().moveGroup(ac.getSession(),
+					ac.getGroupById(parms.readInt("destinationGroup")),
```
Unlike in perun-core, there is no possibility to provide ``null`` for ``destinationGroup``, since it always retrieve group by provided ID. Please make ``destinationGroup`` param optional (check it's presence) to allow moving groups to top-level.

balcirakpeter commented:
Updated javadoc for processRelationMembers().
Fixed method call() in rpc and method moveGroup() in GroupsManagerEntry.

stavamichal commented:
Need to be merged after #1696
